software/statemachine.py

```diff
@@ -0,0 +1,220 @@
+
+        self.robot.move(robotSpeeds['X'], robotSpeeds['Y'], robotSpeeds["ROT"])
+
+    def orbit(self):
+        try:
+            ballCoords = self.getLargestBallCoords(self.imageData)
+            ballXCoord = ballCoords[0]
+            ballYCoord = ballCoords[1]
+        except:
+            self.currentState = State.FIND_BALL
+            return
+
+        normalizedYDistanceFromBottom = (self.imageHeight - ballYCoord) / self.imageHeight
+
+        if normalizedYDistanceFromBottom > 0.6:
+            self.currentState = State.FIND_BALL
+            return
+
+        normalizedXDistanceFromCenter = (ballXCoord - (self.imageWidth / 2)) / self.imageWidth
+        basketMaxDistanceFromCenter = 30 #in pixels
+
+        if self.throwIntoBlue:
+            if abs(self.imageData.basket_b.x - self.imageWidth / 2) < basketMaxDistanceFromCenter:
+                #if orbit is inaccurate, correct position once more
+                if abs(normalizedXDistanceFromCenter) > 0.15 or normalizedYDistanceFromBottom > 0.4:
+                    self.currentState = State.GO_TO_BALL
+                    self.orbitDone = True
+                    return
+                else:
+                    self.currentState = State.THROW
+                    return
+        else:
+            if abs(self.imageData.basket_m.x - self.imageWidth / 2) < basketMaxDistanceFromCenter:
+                if abs(normalizedXDistanceFromCenter) > 0.15 or normalizedYDistanceFromBottom > 0.4:
+                    self.currentState = State.GO_TO_BALL
+                    self.orbitDone = True
+                    return
+                else:
+                    self.currentState = State.THROW
+                    return
+
+        trajectorySpeed = 0.3
+        
+        ySpeedMultiplier = 0.8
+        ySpeed = -(0.3 - normalizedYDistanceFromBottom) * ySpeedMultiplier
+        
+        rotBase = 0.3
+        rotSpeed = rotBase + normalizedXDistanceFromCenter * rotSpeedMultiplier
+        rotSpeedMultiplier = -4
+        
+
+        #if already see basket
+        #chooses which way to orbit
+        if self.throwIntoBlue:
+            if self.imageData.basket_b.exists:
+                trajectorySpeed = 0.2
+                #if basket is on right side of screen, move clockwise
+                if self.imageData.basket_b.x > (self.imageWidth / 2):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+        else:
+            if self.imageData.basket_m.exists:
+                trajectorySpeed = 0.2
+                if self.imageData.basket_m.x > (self.imageWidth / 2):
+                    trajectorySpeed = -trajectorySpeed
+                    rotSpeed = -rotSpeed
+
+        self.robot.move(trajectorySpeed, 0, rotSpeed)
+
+    #averages pixel distances of a 3x3 square from depth_image
+    def basketDist(self, x, y):
```

Comment:
calculate distance in image processor. There is a place there.

Reply:
Added basket distance calculation in image_processor.py analyze_baskets(). Statemachine now gets the distance as a field of the basket object.